sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1727,7 +1744,7 @@ class Analyzer(override val catalogManager: CatalogManager)
 
       case q: LogicalPlan =>
         logTrace(s"Attempting to resolve ${q.simpleString(conf.maxToStringFields)}")
-        q.mapExpressions(resolveExpressionByPlanChildren(_, q))
+        q.mapExpressions(resolveExpressionByPlanChildren(_, q, withMetadata = true))
```

Comment:
Now this is the only place that can resolve to metadata columns. How does it solve the conflicts? Is it possible that the column we are resolving here can be from metadata columns or from `ResolveMissingReferences`?

Reply:
The output ordering inconsistency was due to the `Sort` rule in `ResolveReferences` resolving with metadata columns (which doesn't fall back to this case), when it should have waited for the `Sort` rule in `ResolveMissingReferences`. We still need to be able to resolve metadata columns somewhere.

Reply:
I got rid of this by cherry-picking over the changes in https://github.com/apache/spark/pull/32017, which this PR is blocked on.